### PR TITLE
solving

### PR DESCRIPTION
**Pull request by Poznanskyi**

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,2 +1,84 @@
...
+        if self.hash_table[index] is None:
+            raise KeyError("Cannot find")
+
+        return self.hash_table[index].value
+
+    def __delitem__(self, key: Hashable) -> None:
+        index = self._calculate_index(key)
+
+        if self._calculate_index[index] is None:
+            raise KeyError("Cannot find")
```
Comment (oliasymak24) [inline app/main.py:69]: Cannot find what?

File: app/main.py [hunk 2]
```diff
@@ -1,2 +1,84 @@
...
+    def resize(self) -> None:
+        old_hash_table = self.hash_table
+
+        self.__init__(self.capacity * CAPACITY_MULTIPLIER)
+
+        for node in old_hash_table:
+            if node is not None:
+                self.__setitem__(node.key, node.value)
+
+    def __setitem__(self, key: float, value: float) -> None:
```
Comment (oliasymak24) [inline app/main.py:46]: Bad annotation for key, check everywhere

**Review by oliasymak24 — CHANGES_REQUESTED** [review]
Try to run tests, it is failed, check  it

File: app/main.py [hunk 3]
```diff
@@ -1,2 +1,67 @@
...
+
+    class Node:
+
+        def __init__(self, key: Hashable, value: Any, hash_key: int) -> None:
+            self.key = key
+            self.value = value
+            self.hash_key = hash_key
+
+    def __init__(self) -> None:
+        self.load_factor = 0.667
```
Comment (Oleksl888) [inline app/main.py]: fraction type would be perfect here to present 2/3

File: app/main.py [hunk 4]
```diff
@@ -1,2 +1,67 @@
...
+        self.number_of_elements = 0
+        self.capacity = 8
+        self.size = 0
+        self.hash_table: list[tuple | None] = [None] * self.capacity
```
Comment (Oleksl888) [inline app/main.py]: check annotation - hash table is not a list of tuples

File: app/main.py [hunk 5]
```diff
@@ -1,2 +1,67 @@
...
+    def resize(self) -> None:
+        self.capacity = self.capacity * 2
+        self.size = 0
+        copy_arr: list = deepcopy(self.hash_table)
+        self.hash_table: list[None | Dictionary.Node] = [
+            None for _ in range(self.capacity)]
+
+        for values in copy_arr:
+            if values is not None:
+                self.__setitem__(values.key, values.value)
```
Comment (Oleksl888) [inline app/main.py]: you can use syntactic sugar interface because you defined magic methods already

File: app/main.py [hunk 6]
```diff
@@ -1,2 +1,67 @@
...
+    def __len__(self) -> int:
+        return self.size
+
+    def __str__(self) -> str:
+        return str(self.hash_table)
+
+    def resize(self) -> None:
+        self.capacity = self.capacity * 2
+        self.size = 0
+        copy_arr: list = deepcopy(self.hash_table)
```
Comment (Oleksl888) [inline app/main.py]: you can create an empty hash table here:
`new_table: list[Dictionary.Node | None] = [None] * self.capacity`
then reset it with values from old table in a loop and then simply reassign self.hash_table to new_table

**Review by Oleksl888 — CHANGES_REQUESTED** [review]

File: app/main.py [hunk 7]
```diff
@@ -1,2 +1,74 @@
...
+
+    def __str__(self) -> str:
+        return str(self.hash_table)
+
+    def resize(self) -> None:
+        self.capacity = self.capacity * 2
+        self.size = 0
+        new_table: list[None | Dictionary.Node] = [None] * self.capacity
+
+        new_table: list[None | Dictionary.Node] = [None] * self.capacity
```
Comment (Oleksl888) [inline app/main.py]: why this is duplicated?

File: app/main.py [hunk 8]
```diff
@@ -1,2 +1,74 @@
...
+
+        old_table = self.hash_table
+        self.hash_table = new_table
+
+        for bucket in old_table:
+            if bucket is not None:
```
Comment (Oleksl888) [inline app/main.py]: here can be simplified - you don't need nested for loop here and many if conditions. Just check if bucket is a Node and set it if yes, else do nothing

File: app/main.py [hunk 9]
```diff
@@ -1,2 +1,74 @@
...
+
+    class Node:
+
+        def __init__(self, key: Hashable, value: Any, hash_key: int) -> None:
+            self.key = key
+            self.value = value
+            self.hash_key = hash_key
+
+    def __init__(self) -> None:
+        self.load_factor = 2 / 3
```
Comment (Oleksl888) [inline app/main.py]: still would make sense to use Fraction here to be more precise

**Review by Oleksl888 — CHANGES_REQUESTED** [review]

File: app/main.py [hunk 10]
```diff
@@ -1,2 +1,72 @@
...
+            self.key = key
+            self.value = value
+            self.hash_key = hash_key
+
+    def __init__(self) -> None:
+        self.load_factor = Fraction(2, 3)
+        self.number_of_elements = 0
+        self.capacity = 8
+        self.size = 0
+        self.hash_table = [None] * self.capacity
```
Comment (Oleksl888) [inline app/main.py:19]: `self.hash_table: list[Dictionary.Node | None] = [None] * self.capacity` - adding annotation here will make type hints available in the code below

File: app/main.py [hunk 11]
```diff
@@ -1,2 +1,72 @@
...
+    class Node:
+
+        def __init__(self, key: Hashable, value: Any, hash_key: int) -> None:
+            self.key = key
+            self.value = value
+            self.hash_key = hash_key
+
+    def __init__(self) -> None:
+        self.load_factor = Fraction(2, 3)
+        self.number_of_elements = 0
```
Comment (Oleksl888) [inline app/main.py:16]: This is not used anywhere

File: app/main.py [hunk 12]
```diff
@@ -1,2 +1,72 @@
...
+    def resize(self) -> None:
+        self.capacity = self.capacity * 2
+        self.size = 0
+        new_table: list[None | Dictionary.Node] = [None] * self.capacity
+
+        old_table = self.hash_table
+        self.hash_table = new_table
+
+        for bucket in old_table:
+            if isinstance(bucket, list):
```
Comment (Oleksl888) [inline app/main.py:59]: Should be more simple - something like this:
```
  for node in self.hash_table:
      if isinstance(node, Dictionary.Node):
          index = self.find_index(node.key)
          new_table[index] = node

**Review by Oleksl888 — APPROVED** [review]
Approved - but please check last comments

**Pull request closed without merging**